### PR TITLE
Split phase change from `MirPass`

### PR DESCRIPTION
**Pull request by JakobDegen**
The main goal here is to simplify the pass manager logic. `MirPass` no longer contains the `phase_change` method, and `run_passes` instead accepts an `Option<PhaseChange>`. The hope is that this addresses the comments (and maybe perf regression) from #99102 .

r? @oli-obk cc @RalfJung

### CODE REVIEW

File: compiler/rustc_middle/src/mir/mod.rs [hunk 1]
```diff
@@ -292,6 +319,7 @@ impl<'tcx> Body<'tcx> {
 
         let mut body = Body {
             phase: MirPhase::Built,
+            pass_count: 1,
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/mod.rs:322]: That looks like pretty fragile initialization. Is this always 1 initially?

Comment (JakobDegen) [reply compiler/rustc_middle/src/mir/mod.rs:322]: Yeah, this is a counter that is always set/reset to 1 everywhere.

**Comment by JakobDegen** [comment]
The revert is entirely a shot in the dark for getting the perf from #99102 back, but let's see (waiting on CI before queuing)

**Comment by JakobDegen** [comment]
I've dropped the last commit (which turned out to not give back any perf benefits). I'll keep looking into fixing the perf regression from the other PR, but in the meantime I think we should probably merge this, since it's an improvement on its own.

**Comment by JakobDegen** [comment]
Was the failure spurious? How do I tell?

**Comment by oli-obk** [comment]
@bors retry timeout

Probably just an arbitrary spurious CI failure

**Comment by Mark-Simulacrum** [comment]
FYI using this commit to get a SMT + Turbo vs. non-SMT and non-Turbo comparison on the perf machine, expecting another post here as a result. See discussion here https://rust-lang.zulipchat.com/#narrow/stream/247081-t-compiler.2Fperformance/topic/Disabling.20hyperthreading.2Fturbo/near/306100379

**Pull request merged**